django/contrib/messages/storage/cookie.py

```diff
@@ -150,16 +153,23 @@ def _decode(self, data):
         """
         if not data:
             return None
-        bits = data.split('$', 1)
-        if len(bits) == 2:
-            hash, value = bits
-            if constant_time_compare(hash, self._hash(value)):
-                try:
-                    # If we get here (and the JSON decode works), everything is
-                    # good. In any other case, drop back and return None.
-                    return json.loads(value, cls=MessageDecoder)
-                except json.JSONDecodeError:
-                    pass
+        bits = data.split('$', 2)
+        if len(bits) >= 2:
+            # Before Django 3.1, no algorithm prefix was set.
+            hash_algorithm = bits[0] if len(bits) == 3 else 'sha1'
```

Comment:
There should be a whitelist of supported algorithms. Since the user has direct access to the cookie, this makes it vulnerable to downgrade attacks. It can force a fallback to MD5, or maybe even worse on particular Python/OpenSSL installations.

Reply:
What Marten said, this is user controlled data and reminds me of the JWT issues with setting algo to none :D Whitelist seems like an option, paging @ubernostrum if he has even better ideas.

Reply:
No, we should have a curated (minimal) list of what we want to support, external changes due to a python update (ie in `hashlib.algorithms_available`) are not okay imo.

Reply:
The question is if we expect/want users to change this, if not I'd opt for sha1,blake2b and then drop sha1 as needed. With security imo less options is better.

I'd also wait for more feedback from people with a bit more knowledge on how common or useful it is to use blake with hmacs. In the end blake afaik has it's own hmac like mode -- not sure if that is good or bad though :)